### ISSUE TITLE
Square root of negative numbers in heaters.py

### ISSUE THREAD

**Issue by Duckle29** [KevinOConnor/klipper#245]
Hello there. Depending on the parameters for the heaters, you can run into a scenario where you're squaring negative numbers, causing a `ValueError: math domain error`

Here's the parameters for the thermistor giving issues:

```python
# Available sensors
Sensors = {
	"AliRubber": {
		'class': Thermistor, 't1': 25., 'r1': 15633.,
                't2': 75., 'r2': 12425., 't3': 125., 'r3': 6852. },
}
```

The issues are with the calculation of x. With this particular set of values, c2 becomes negative, and that means you'll take a square root of a negative number, and break into the complex domain.

I've attached a PDF of the values run through in maple:
[root of negative.pdf](https://github.com/KevinOConnor/klipper/files/1822101/root.of.negative.pdf)

If the sign of the value in the root doesn't matter, a fix could be the following:

```python
x = math.sqrt(math.abs(((self.c2 / (3. * self.c3))**3 + y**2)))
```

**Comment by KevinOConnor** [comment]
Thanks.  In a conversation on IRC, I'm pretty sure we determined that the "AliRubber" uses a standard 100K beta 1950 thermistor.  So, no new changes are needed to support it.

It is possible to change the math to allow more values, but I don't think that is necessary for real thermistors.  A real thermistor has positive values for c1, c2, and c3 that wouldn't result in a negative value in this caluclation.  So, increasing the range just allows the error to propagate further.

-Kevin

**Comment by Duckle29** [comment]
That's a valid point. Maybe alter the error message a bit, so that it says "Are you sure your R/T pairs are correct?"

**Comment by KevinOConnor** [comment]
FYI, this came up again on IRC.  It does seem like we'll need to do something about this once we add support for adding custom thermistors to the config file.  (Either allowing c3 to be negative or redoing the calculations with c3=0 if it is found to be negative.)  In any case, this will be something to look at after the v0.6.0 release.

**Comment by Duckle29** [comment]
Another option would be to just use Steinhart hart coefficients directly, and ask people to use one of the many calculators online? 

Alternatively here's how Smoothieware does it: https://github.com/Smoothieware/Smoothieware/blob/45d39d391b8e50f2875197fca23e0a97d5397222/src/modules/utils/simpleshell/SimpleShell.cpp#L876

**Comment by KevinOConnor** [comment]
I was able to backtrack the math in that calculation (`math.sqrt(((self.c2 / (3. * self.c3))**3 + y**2))`).  It's using "Cardano's formula" for solving a cubic formula.  Unfortunately, when the value is negative, it means there are multiple real roots for the cubic equation.

I don't think the c3 coefficient should be negative in practice on Steinhart-Hart - I suspect it would only be negative if the c3 is actually very small and there is some noise in the measurements.  So, I don't think it makes sense to complicate the math to handle this corner case.

So, I updated the code to fallback to a simple "beta" style calculation in this case (commit 75b89f52).
